### PR TITLE
UPDATE: Rubber, Game, Trick Basemodel

### PR DESCRIPTION
**Pull request by Segelzwerg**

### CODE REVIEW

**Comment by Segelzwerg** [comment]
There seems to be an issue with https://pydantic-docs.helpmanual.io/usage/exporting_models/#json_encoders

**Comment by Segelzwerg** [comment]
Coverage test will follow, but please review principal of PR.

File: whist/core/game/hand.py [hunk 1]
```diff
@@ -10,38 +12,35 @@
 from whist.core.game.warnings import TrickNotDoneWarning
 
 
-class Hand:
+class Hand(BaseModel):
     """
     Hand of whist.
     """
+    tricks: list[Trick] = []
+    trump: Suit = None
```
Comment (iTitus) [inline whist/core/game/hand.py:20]: Optional[Suit] ?

File: whist/core/game/game.py [hunk 2]
```diff
@@ -1,25 +1,35 @@
...
+    current_hand: Optional[Hand] = None
+
+    # pylint: disable=too-few-public-methods
+    class Config:
+        """
+        Enables to have non models as field types and sets the encoder for play order.
+        """
+        arbitrary_types_allowed = True
+        json_encoders = {
+            PlayOrder: lambda x: json.dumps(x, cls=PlayOrder.PlayOrderEncoder)
```
Comment (iTitus) [inline whist/core/game/game.py:31]: this converts the PlayOrder to a JSON string for saving -> https://pydantic-docs.helpmanual.io/usage/exporting_models/#json_encoders

Comment (Segelzwerg) [reply whist/core/game/game.py:31]: Yes, you want me to add a comment?

File: whist/core/game/hand.py (hunk 1 above)
Comment (Segelzwerg) [reply whist/core/game/hand.py:20]: Yes

File: whist/core/game/game.py (hunk 2 above)
Comment (iTitus) [reply whist/core/game/game.py]: Thanks I hate it

File: whist/core/util.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    """
+    Forces a dictionary to use string values instead of objects for the given keys.
+    :param dictionary: which needs transformation
+    :param keys: that values needed to be transformed
+    :return: the above dictionary with values changes to string
+    """
+    for key in keys:
+        if key in dictionary:
+            dictionary[key] = str(dictionary[key]) if dictionary[key] else None
+    return dictionary
```
Comment (iTitus) [inline whist/core/util.py:16]: change type hint for arg from set to iterable or something.

File: whist/core/game/trick.py [hunk 4]
```diff
@@ -60,3 +61,7 @@ def play_card(self, player: PlayerAtTable, card: Card) -> None:
             raise ServSuitFirstWarning()
 
         self.stack.add(card)
+
+    def dict(self, *args, **kwargs):
+        super_dict = super().dict(*args, **kwargs)
+        return enforce_str_on_dict(super_dict, {'trump'})
```
Comment (iTitus) [inline whist/core/game/trick.py:67]: this doesnt need to be a set

File: whist/core/game/play_order.py [hunk 5]
```diff
@@ -1,85 +1,83 @@
...
     def get_player(self, player: Player) -> PlayerAtTable:
         """
-        Retrieves the PlayerAtTable for the player given.
-        :param player: who needs it's counterpart at the table
-        :return: the player at table
-        """
+            Retrieves the PlayerAtTable for the player given.
+            :param player: who needs it's counterpart at the table
+            :return: the player at table
+            """
```
Comment (iTitus) [inline whist/core/game/play_order.py:72]: this diff looks strange, is the indentation correct?

File: whist/core/game/hand.py [hunk 6]
```diff
@@ -62,13 +62,17 @@ def next_trick(self, play_order: PlayOrder) -> Trick:
...
         next_trick_order = self._winner_plays_first_card(play_order)
-        next_trick = Trick(play_order=list(next_trick_order), trump=self._trump)
-        self._tricks.append(next_trick)
+        next_trick = Trick(play_order=list(next_trick_order), trump=self.trump)
+        self.tricks.append(next_trick)
         return next_trick
 
+    def dict(self, *args, **kwargs):
+        super_dict = super().dict(*args, **kwargs)
+        return enforce_str_on_dict(super_dict, {'trump'})
```
Comment (iTitus) [inline whist/core/game/hand.py:74]: this doesnt need to be a set

File: whist/core/game/play_order.py [hunk 7]
```diff
@@ -1,85 +1,83 @@
...
         """
-        Rotates the play order, so the player will be next player.
-        :param player: who should be at beginning of the play order
-        :return: None
-        """
+            Rotates the play order, so the player will be next player.
+            :param player: who should be at beginning of the play order
+            :return: None
+            """
         order = list(self)
```
Comment (iTitus) [inline whist/core/game/play_order.py:47]: indent?

File: whist/core/game/play_order.py [hunk 8]
```diff
@@ -1,85 +1,83 @@
...
 
     def next_order(self) -> 'PlayOrder':
         """
-        Create the order for the next hand.
-        :rtype: PlayOrder
-        """
-        return PlayOrder._new_order(self)
+            Create the order for the next hand.
+            :rtype: PlayOrder
+            """
```
Comment (iTitus) [inline whist/core/game/play_order.py:55]: indent?

**Review by iTitus — CHANGES_REQUESTED** [review]

File: whist/core/util.py (hunk 3 above)
Comment (Segelzwerg) [reply whist/core/util.py:16]: Just following pydantic standards which uses sets

**Review by iTitus — APPROVED** [review]

**Pull request merged**